src/Agent.Listener/Configuration/SystemdControlManager.cs

```diff
@@ -0,0 +1,148 @@
+using System.IO;
+using System.Linq;
+using System.Text;
+
+using Microsoft.VisualStudio.Services.Agent.Util;
+using System.Threading;
+using System.Diagnostics;
+
+namespace Microsoft.VisualStudio.Services.Agent.Listener.Configuration
+{
+    public class SystemDControlManager : ServiceControlManager
+    {
+        // This is the name you would see when you do `systemctl list-units | grep vsts`
+        private const string ServiceNamePattern = "vsts.agent.{0}.{1}.service";
+        private const string ServiceDisplayNamePattern = "VSTS Agent ({0}.{1})";
+
+        private const int MaxUserNameLength = 32;
+        private const string VstsAgentServiceTemplate = "vsts.agent.service.template";
+        private const string _shTemplate = "systemd.svc.sh.template";
+        private const string _shName = "svc.sh";
+
+        public override bool ConfigureService(
+            AgentSettings settings,
+            CommandSettings command)
+        {
+            Trace.Entering();
+
+            var _linuxServiceHelper = HostContext.GetService<INativeLinuxServiceHelper>();
+            CalculateServiceName(settings, ServiceNamePattern, ServiceDisplayNamePattern);
+
+            if (!_linuxServiceHelper.CheckIfSystemdExists())
+            {
+                Trace.Info("Systemd does not exists, returning");
+                _term.WriteLine(StringUtil.Loc("SystemdDoesNotExists"));
+
+                return false;
+            }
+
+            var unitFile = _linuxServiceHelper.GetUnitFile(settings.ServiceName);
+
+            if (CheckServiceExists(settings.ServiceName))
+            {
+                _term.WriteError(StringUtil.Loc("ServiceAlreadyExists", unitFile));
+                throw new InvalidOperationException(StringUtil.Loc("CanNotInstallService"));
+            }
+
+            try
+            {
+                string svcShPath = Path.Combine(IOUtil.GetRootPath(), _shName);
+
+                string svcShContent = File.ReadAllText(Path.Combine(IOUtil.GetBinPath(), _shTemplate));
+                var tokensToReplace = new Dictionary<string, string>
+                                          {
+                                              { "{{SvcDescription}}", settings.ServiceDisplayName },
+                                              { "{{SvcNameVar}}", settings.ServiceName }
+                                          };
+
+                svcShContent = tokensToReplace.Aggregate(
+                    svcShContent,
+                    (current, item) => current.Replace(item.Key, item.Value));
+
+                File.WriteAllText(svcShPath, svcShContent, Encoding.UTF8);
+
+                var unixUtil = HostContext.CreateService<IUnixUtil>();
+                unixUtil.Chmod("755", svcShPath).GetAwaiter().GetResult();
+
+                SvcSh("install");
+                _term.WriteLine(StringUtil.Loc("ServiceConfigured", settings.ServiceName));
+            }
+            catch (UnauthorizedAccessException ex)
+            {
```

Comment:
this exception is no longer thrown (it worked with pinvoke commands before). i need to find out how to handle permission errors